### DISCUSSION TITLE
How to hardcode the database schema

### DISCUSSION THREAD

**Discussion by MarvinJWendt** [pocketbase/pocketbase#1052] [Q&A] [answered]
Hi, we want to publish a server (as a docker image). The backend of that server will be PocketBase as a framework, and it will be compiled into the same binary that also serves the frontend (via the PocketBase router).

As the users use the docker image, they will not be able to change our backend or frontend code. We will also disable editing the collection schema by setting [`MetaConfig.HideControls`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.7.10/core#MetaConfig.HideControls) by default. 

Now, I want to initzialize PocketBase with a predefined schema, because the user cannot (and should not need to) edit it. 

How is that possible?

```mermaid
graph TD
    A[server startup] --> B{has collections}
    B -->|true| FIN
    B -->|false| C[Create collections from schema json]
    C --> FIN

    FIN[server running]
```

**Comment by ganigeorgiev** [comment, marked as answer]
If you are using PocketBase as framework, you can create a [DB migration](https://pocketbase.io/docs/db-migrations/).

Migrations are automatically executed when running the `serve` command.

The migrations are go files so they will be part of your executable and you can technically put whatever you want in them that needs to be executed only once.
There will be more documentation how to create/update a collection programmatically and how to manage migrations in the upcoming v0.8.0 release later this week.

**Reply by ganigeorgiev** [reply]
Also please note that `HideControls` setting only hides the collection modify controls from the UI, but if you have an admin account then you'll still be able to change your collection via the API. All admins act as "dev/super-admin" accounts and they have access to everything (collection API rules are ignored).

**Reply by MarvinJWendt** [reply]
Oh nice, thank you. I was aware about the migrations, but I didn't know they could be generated, which makes it a lot easier. Just out of interest, is it possible to call the method to import collections from an exported schema JSON manually in the go code?

**Reply by MarvinJWendt** [reply]
> but if you have an admin account then you'll still be able to change your collection via the API

That's fine, if they really try to mess their server up, I'll let them do it :) I just want to avoid accidental schema updates.

**Reply by ganigeorgiev** [reply]
> is it possible to call the method to import collections from an exported schema JSON manually in the go code?

Yes. That actually is how `migrate collections` works - it creates a file that calls internally `dao.ImportCollection(collections, true, nil)`, where `collections` is a json array of collections - https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.8.0-rc3/daos#Dao.ImportCollections

**Reply by ganigeorgiev** [reply]
btw. you can also prevent collection changes over the `/api/collection` APIs by attaching to the collection request hooks and return an error, eg. something like:

```go
app.OnCollectionBeforeCreateRequest().Add(func(e *core.CollectionCreateEvent) error {
    return apis.NewForbiddenError("Collection create is forbidden.", nil)
})

app.OnCollectionBeforeUpdateRequest().Add(func(e *core.CollectionUpdateEvent) error {
    return apis.NewForbiddenError("Collection update is forbidden.", nil)
})

app.OnCollectionBeforeDeleteRequest().Add(func(e *core.CollectionDeleteEvent) error {
    return apis.NewForbiddenError("Collection delete is forbidden.", nil)
})

```
